common/src/main/java/io/netty/util/Recycler.java

```diff
@@ -190,557 +162,149 @@ public final boolean recycle(T o, Handle<T> handle) {
             this.parent = parent;
-            threadRef = new WeakReference<Thread>(thread);
             this.maxCapacity = maxCapacity;
-            availableSharedCapacity = new AtomicInteger(max(maxCapacity / maxSharedCapacityFactor, LINK_CAPACITY));
-            elements = new DefaultHandle[min(INITIAL_CAPACITY, maxCapacity)];
-            this.interval = interval;
-            this.delayedQueueInterval = delayedQueueInterval;
-            handleRecycleCount = interval; // Start at interval so the first one will be recycled.
-            this.maxDelayedQueues = maxDelayedQueues;
-        }
-
-        // Marked as synchronized to ensure this is serialized.
-        synchronized void setHead(WeakOrderQueue queue) {
-            queue.setNext(head);
-            head = queue;
-        }
-
-        int increaseCapacity(int expectedCapacity) {
-            int newCapacity = elements.length;
-            int maxCapacity = this.maxCapacity;
-            do {
-                newCapacity <<= 1;
-            } while (newCapacity < expectedCapacity && newCapacity < maxCapacity);
-
-            newCapacity = min(newCapacity, maxCapacity);
-            if (newCapacity != elements.length) {
-                elements = Arrays.copyOf(elements, newCapacity);
-            }
-
-            return newCapacity;
-        }
-
-        @SuppressWarnings({ "unchecked", "rawtypes" })
-        DefaultHandle<T> pop() {
-            int size = this.size;
-            if (size == 0) {
-                if (!scavenge()) {
-                    return null;
-                }
-                size = this.size;
-                if (size <= 0) {
-                    // double check, avoid races
-                    return null;
-                }
-            }
-            size --;
-            DefaultHandle ret = elements[size];
-            elements[size] = null;
-            // As we already set the element[size] to null we also need to store the updated size before we do
-            // any validation. Otherwise we may see a null value when later try to pop again without a new element
-            // added before.
-            this.size = size;
-
-            if (ret.lastRecycledId != ret.recycleId) {
-                throw new IllegalStateException("recycled multiple times");
-            }
-            ret.recycleId = 0;
-            ret.lastRecycledId = 0;
-            return ret;
+            this.ratioInterval = ratioInterval;
+            pooledHandles = new ArrayDeque<DefaultHandle<T>>();
+            externalInbox = new ArrayDeque<DefaultHandle<T>>();
+            ratioCounter = ratioInterval; // Start at interval so the first one will be recycled.
         }
 
-        private boolean scavenge() {
-            // continue an existing scavenge, if any
-            if (scavengeSome()) {
-                return true;
+        DefaultHandle<T> claim() {
+            if (flagInbox) {
+                drainInbox();
```

Comment:
draining inbox would worth if there isn't anything on pooledHandles so we can defer draining until is necessary or if the external mailbox got "too big"?

Reply:
No longer relevant.